### PR TITLE
Streamer uses KafkaW

### PR DESCRIPTION
**Pull request by michele-brambilla**
This PR introduces two main changes

1. the ``streamer``now  uses ``KafkaW`` instead of librdkafka; 
2. the ``connect()`` task is now executed in task-based parallelism instead of thread-based;

and a new feature

3. added tests for the ``Streamer``.

Consequences:

1) makes the filewriter implementation more uniform. ``KafkaW`` is the only explicit connection with librdkafka;
2) let us remove explicit locks and condition variables. The flow is now more linear;
3) needs the fake librdkafka implementation to be specified using ``-DFAKE_RDKAFKA=<path>`` when invoking ``cmake``.

### CODE REVIEW

**Comment by michele-brambilla** [comment]
Please note that some cleanup is still required. Nevertheless I would like to have your opinion and any suggestion.

File: src/Msg.h [hunk 1]
```diff
@@ -77,7 +77,7 @@ class Msg {
     return msg;
   }
 
-  static Msg fromKafkaW(std::unique_ptr<KafkaW::Msg> KafkaWMsg) {
+  static Msg fromKafkaW(std::unique_ptr<KafkaW::Msg> &&KafkaWMsg) {
```
Comment (matthew-d-jones) [inline src/Msg.h]: It looks to me like `fromKafkaW` should take ownership of the the `KafkaWMsg` pointer, and therefore passing it by value makes sense to me, why the change?

File: src/Streamer.cxx [hunk 2]
```diff
@@ -206,28 +218,31 @@ void FileWriter::Streamer::connect(const std::string &TopicName) {
...
-    return;
-  }
-  LOG(Sev::Debug, "createTopicPartition");
-  RunStatus = assignTopicPartition();
-  if (RunStatus != SEC::no_error) {
-    return;
-  }
+  ConsumerW.reset(new KafkaW::Consumer(Settings));
+  ConsumerW->addTopic(TopicName);
+  ////
```
Comment (matthew-d-jones) [inline src/Streamer.cxx]: Don't forget to clean up this

File: src/Msg.h (hunk 1 above)
Comment (michele-brambilla) [reply src/Msg.h]: You're right, that should not have changed

**Comment by michele-brambilla** [comment]
I have no more commit to push, any review and suggestion is welcome. I modified the initial message to give a better overview of the changes.

File: src/Streamer.cxx [hunk 3]
```diff
@@ -1,321 +1,143 @@
...
-    return ProcessMessageResult::OK();
+  // wait for connect() to finish
+  try {
+    if (IsConnected.valid()) {
+      if (IsConnected.wait_for(std::chrono::milliseconds(1000)) !=
+          std::future_status::ready) {
+        LOG(Sev::Critical, "... still not ready");
+        return ProcessMessageResult::OK();
+      }
+      std::call_once(ConnectionStatus,
```
Comment (dominikwerder) [inline src/Streamer.cxx]: Why is `std::call_once` necessary here?

File: src/tests/CMakeLists.txt [hunk 4]
```diff
@@ -15,18 +16,33 @@ status_test.cxx
 status_writer.cxx
 command_handler.cxx
 stream_master.cxx
-MainOpt_tests.cxx
```
Comment (dominikwerder) [inline src/tests/CMakeLists.txt]: Why are they removed?

File: src/Streamer.cxx (hunk 3 above)
Comment (michele-brambilla) [reply src/Streamer.cxx]: Not anymore, before the fture::get was outside of the if condition. Fixed.

File: src/tests/CMakeLists.txt (hunk 4 above)
Comment (michele-brambilla) [reply src/tests/CMakeLists.txt]: No reason, restored.

File: src/StreamMaster.h [hunk 5]
```diff
@@ -45,9 +45,13 @@ template <typename Streamer> class StreamMaster {
...
-                        std::forward_as_tuple(d.topic()),
-                        std::forward_as_tuple(broker, d.topic(), options));
+      try {
+        Streamers.emplace(std::piecewise_construct,
+                          std::forward_as_tuple(d.topic()),
+                          std::forward_as_tuple(broker, d.topic(), options));
+      } catch (std::exception &e) {
+        LOG(Sev::Critical, "{}", e.what());
+      }
       Streamers[d.topic()].setSources(d.sources());
```
Comment (dominikwerder) [inline src/StreamMaster.h]: Does this behave correctly if the code in `try` throws?

File: src/Streamer.cxx [hunk 6]
```diff
@@ -1,321 +1,142 @@
...
+    }
+  } catch (std::exception &Error) {
+    LOG(Sev::Critical, "{}", Error.what());
   }
-  std::lock_guard<std::mutex> lock(
-      ConnectionReady); // make sure that connect is completed
 
+  // make sure that the connection is ok
+  // attention: connect() handles exceptions
   if (int(RunStatus) < 0) {
```
Comment (dominikwerder) [inline src/Streamer.cxx]: Instead of C style cast, is it possible to have some `RunStatus.connectionOk()` ?

File: src/tests/streamer_options.cxx [hunk 7]
```diff
@@ -57,52 +62,38 @@ void StreamerOptionsTest::CompareSteamerOptionsWith(
...
-  rapidjson::Document optj;
-  optj.Parse(R""({
-        "streamer" : {
-            }
-      })"");
-  ConfigureStreamerWithJson(optj);
-  FileWriter::StreamerOptions Defaults;
-  CompareSteamerOptionsWith(Defaults);
-}
+// TEST_F(StreamerOptionsTest, streamer_empty_json_doesn_t_change_defaults) {
```
Comment (dominikwerder) [inline src/tests/streamer_options.cxx]: Can this commented test be removed?

File: src/CommandHandler.cxx [hunk 8]
```diff
@@ -353,6 +351,8 @@ void CommandHandler::handleStreamMasterStop(std::string const &Command) {
   if (!MasterPtr) {
     return;
   }
+  LOG(Sev::Critical, "{}", Command);
```
[the remark this answers is not in the document]
Comment (michele-brambilla) [reply src/CommandHandler.cxx]: Is there some other place where we can log (maybe in Debug mode only) the command? It could be helpful for debugging. If not, I would suggest to keep this log and set it to Sev::Debug.

File: src/Streamer.cxx (hunk 6 above)
Comment (michele-brambilla) [reply src/Streamer.cxx]: The two ``enums`` (StreamerErrorCode and StreamMasterErrorCode) are replaced with two classes. Now it is possible.

File: src/tests/streamer_options.cxx (hunk 7 above)
Comment (michele-brambilla) [reply src/tests/streamer_options.cxx]: :+1:

File: src/StreamMaster.h (hunk 5 above)
Comment (michele-brambilla) [reply src/StreamMaster.h]: The behaviour was in any case correct, nevertheless it is more clean as we discuss this morning.
